### ISSUE TITLE
Check for available resources before starting instances in jobs

### ISSUE THREAD

**Issue by zxiiro** [jenkinsci/openstack-cloud-plugin#89]
It would be nice if jobs that declare they need additional resources can check that there is enough resources in openstack before trying to spin up VMs. This feature should also allow retrying up to a timeout limit before failing the job.

Our use case is that we have many integration jobs that spin up full labs of VMs before running the job. The current openstack plugin fails the build if the instances could not be spun up. We would prefer that the job confirmed that there was enough resources to spin up the additional instances before trying, and if the resources were not available check again periodically until I can allocate the resources. It should then fail the build only once reaching some timeout value like 30 minutes or so.

**Comment by olivergondza** [comment]
I am not sure how to proceed here. I intent to introduce quotas inspection into the plugin so it will not try to provision when not sufficient. My primal concern is slave provisioning but I guess it can be extended to per-build server provisioning. Are you using that to prvision your resources?

In case you do not, we will have to expose quotas via REST or some pipeline API.

**Comment by zxiiro** [comment]
I don't know much about Heat but I've been informed that it can be used for handling this kind of problem and that ideally that should be the implementation so I opened #88 to discuss that possibility. This bug is open in case adding Heat might be complicated and we might want to investigate something similar in the short term.

What we're using is the openstack-cloud-plugin configuration at the job level. So a job starts up and using openstack-cloud-plugin it wants to spin up additional lab instances before running. Currently if openstack is unable to spin up the instances for whatever reason, it immediately fails the build which we don't like. We would rather be able to configure a retry/timeout limit and allow openstack-cloud-plugin to keep checking / trying until it is able to allocate the resources.

JClouds plugin handled this by retrying a few times before giving up but we've noticed with OpenStack plugin it tries a single time and immediately fails our jobs. This is a problem for us because we often expect a large queue of jobs that need to run in our CI but since switching to openstack plugin our queue is parsed very quickly if there is an issue such as out of resources and all jobs fail.

**Comment by olivergondza** [comment]
Just to clarify, you are using *OpenStack Instance Creation* that fails in case the provisioning is not successful on the first try. I guess we can do some magic to tell likely persistent (incorrect configuration) and likely temporary (openstack failure, quota) problems apart and implement a timeout to wait for in case we believe further attempts are likely to succeed.

---

Btw, will https://github.com/jenkinsci/openstack-cloud-plugin/issues/87#issuecomment-226731749 satisfy this use-case? (The effective difference would be the machine provisioned will be the actual slave and not additional server)

**Comment by zxiiro** [comment]
Yes, we use the _OpenStack Instance Creation_ checkbox in a job. The behaviour we noticed is that it tries to start the defined instances but if any failures happen for any reason it immediately fails the job and destroys whatever slaves it did spin up. The frustration with this issue is that let's say you have a job queue of 20 jobs that all use the _OpenStack Instance Creation_ and the overall system has reached it's quota, all 20 jobs will immediately fail dropping the queue to 0 rather than continuing to stay queued until resources are available again.

I'm not sure if the suggested comment will satisfy the use-case as I don't fully understand it. But if it is able to spin up the full set of systems defined in the template before starting the job ensuring that jobs don't fail just because of resources are not available then I think that should be fine with us. Our main concern with the current implementation is that our jobs all instantly fail instead of staying in the queue until it is possible to run the jobs.
